### PR TITLE
flac: add support for LPC encoding

### LINKED ISSUE #35: Implement LPC encoding of audio samples
Notes from @wsc1 at https://github.com/mewkiz/flac/pull/32#issuecomment-420835961

> For the LPC stuff, just wanted to note
> [this mail](http://lists.xiph.org/pipermail/flac-dev/2011-July/002999.html)
> 
> [this chapter](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.113.8455&rep=rep1&type=pdf)
> 
> and [zc lpc](http://godoc.org/zikichombo.org/dsp/lpc)
> 
> So one possible way to approach flac LPC (the fixed LPC shouldn't require this BTW) would be 
> to extend zc/lpc to use line spectral pairs, expose the parcor coefficients, and work with some of the quantisation methods in the chapter.
> 
> If that is of interest to you, I'm happy to collaborate.

### PR DESCRIPTION
**Pull request by mewmew**
Fixes #35

### CODE REVIEW

**Comment by mewmew** [comment]
This PR adds support for Fixed and FIR LPC encoding of audio samples.

A set of [round-trip test cases](https://github.com/mewkiz/flac/blob/2696ca5ca796c398a4c9fef7ca34107d549a4684/enc_test.go#L13) are used to ensure that decoding/encoding result in bit-by-bit identical FLAC files.

While many test cases pass, there are also a lot of test cases that produce diffs, and these should be investigated before merging this PR.

```bash
go test -v -run TestEncode
```

Furthermore, there are [two](https://github.com/mewkiz/flac/blob/2696ca5ca796c398a4c9fef7ca34107d549a4684/encode_frame.go#L68) [TODOs](https://github.com/mewkiz/flac/blob/2696ca5ca796c398a4c9fef7ca34107d549a4684/encode_subframe.go#L25) that should be discussed to ensure that the act of encoding audio samples is non-destructive (i.e. does not modify `frame.Subframes[i].Samples`.

**Comment by mewmew** [comment]
The [fq](https://github.com/wader/fq) tool by @wader has been invaluable to investigate the cause of diffs between input (decoding) and output (encoding) FLAC files.

To investigate a mismatch between input FLAC files and corresponding FLAC files produced by encode, do as follows.

1. [uncomment the `ioutil.WriteFile` lines](https://github.com/mewkiz/flac/blob/3a6ac918b27457e9b2df73a2c60256b7fd3b5f26/enc_test.go#L203) in `TestEncode` to output FLAC file (to `/tmp/got.flac`) when the contents of the round-trip test case mismatch.
2. [uncomment a test case that is currently failing](https://github.com/mewkiz/flac/blob/3a6ac918b27457e9b2df73a2c60256b7fd3b5f26/enc_test.go#L27) (e.g. `testdata/59996.flac`)
3. run `go test -v -run TestEncode/testdata/59996.flac`
```
=== RUN   TestEncode/testdata/59996.flac
    enc_test.go:206: "testdata/59996.flac": content mismatch;
    expected 66 4C 61 ... 7C A2 21 22 3B ... 80 32 2A,
    got      66 4C 61 ... 7C 62 21 12 3B ... 80 32 2A
                             ^     ^
                             |     |
                             |     |
                            DIFF  DIFF
--- FAIL: TestEncode (0.02s)
    --- FAIL: TestEncode/testdata/59996.flac (0.02s)
```
4. use [fq](https://github.com/wader/fq) to analyze the diff
   - 4a: analyze input FILE (source of truth): `fq d -d flac testdata/59996.flac`
   - 4b: analyze output FILE (produced by encode): `fq d -d flac /tmp/got.flac`

## binary diff (using vbindiff) between input and output FLAC files

screenshot_2023-10-29_11:10:27

## fq analysis of input FILE (source of truth)

```bash
fq d -d flac testdata/59996.flac
```

screenshot_2023-10-29_11:14:40

## fq analysis of output FILE (produce by `mewkiz/flac` encoding)

```bash
fq d -d flac /tmp/got.flac
```

screenshot_2023-10-29_11:14:50

**Comment by wader** [comment]
🥳  happy to hear that!

Some additional tricks i usually use with fq:
- Do "structural" diff. Will output a JSON structure with differences: `fq -n 'diff(input; input)' a b`. Does not work great when arrays are diffing.
- Use same query for two or more files `fq 'first(grep_by(.type=="stts"))' a.mp4 b.mp4`
- Combined `fq -n 'def f: first(grep_by(.type=="stts")); diff(input|f; input|f)' a.mp4 b.mp4`
- `dd` is nice if you want to see all bytes, `ddv` if you want to see more start/stop offset details

BTW the flac decoder in fq does sample decoding to be able to verify md5 but does not expose the sampels in the decode tree as it would use a bit too much memory for most flac files. But i could probably add an option to do if it would be useful? 

Also nice to see more encoding support! i've thought learning more about how that is done but all the LPC stuff feels a bit overwhelming. Will surely be following this PR with great interest!

**Comment by mewmew** [comment]
> BTW the flac decoder in fq does sample decoding to be able to verify md5 but does not expose the sampels in the decode tree as it would use a bit too much memory for most flac files. But i could probably add an option to do if it would be useful?

Having an experimental branch of fq that outputs samples (and residuals if possible) would definitely be useful for trouble-shooting : )

**Comment by mewmew** [comment]
```diff
--input (orig)
++output (encode)
 residual: (0b-0000000000000000011111100101101) -16173
 residual: (0b-0000000000000000011101101010011) -15187
 residual: (0b00000000000000000111101011001001) 31433
-residual: (0b-0000000000000000001110111110011) -7667
+residual: (0b-0000000000000000001110111110010) -7666
-residual: (0b00000000000000000110001000010010) 25106
+residual: (0b00000000000000000110001000010001) 25105
-residual: (0b00000000000000000010001110110100) 9140
+residual: (0b00000000000000000010001110110101) 9141
-residual: (0b-0000000000000000110000000101110) -24622
+residual: (0b-0000000000000000110000000110000) -24624
-residual: (0b00000000000000000001000011000000) 4288
+residual: (0b00000000000000000001000011000010) 4290
-residual: (0b00000000000000000110101101001001) 27465
+residual: (0b00000000000000000110101101000110) 27462
-residual: (0b-0000000000000000111100011100001) -30945
+residual: (0b-0000000000000000111100011011110) -30942
-residual: (0b00000000000000000111111110100011) 32675
+residual: (0b00000000000000000111111110100000) 32672
-residual: (0b-0000000000000000101000111001011) -20939
+residual: (0b-0000000000000000101000111001001) -20937
-residual: (0b-0000000000000000011011001011101) -13917
+residual: (0b-0000000000000000011011001011110) -13918
 residual: (0b00000000000000001000100101110011) 35187
 residual: (0b-0000000000000001101011111001001) -55241
```

There seem to be a small diff in the residual between the original input FLAC file and the encoded output FLAC file.

**Comment by wader** [comment]
@mewmew Did a quick PR https://github.com/wader/fq/pull/802 run with `-o sample_details=true`. This also reminded me that maybe some usages of "sample" in the current output should actually be "residual"?

**Comment by wader** [comment]
> Still not sure why. Anyone who may have an idea? : )

What residual encoding mode is this? rice parameter or escape? some interesting diff in high/low or before/after zigzag? all from same partition?

**Comment by mewmew** [comment]
> What residual encoding mode is this? rice parameter or escape? some interesting diff in high/low or before/after zigzag? all from same partition?

The residuals are encoded using using rice parameter. There could for sure be issues in the high/low and/or zigzagging parts of encoding. Another aspect I think may be at fault (I think there may be more than one that's slightly broken) is the [inter-channel decorrelation](https://github.com/mewkiz/flac/blob/3a6ac918b27457e9b2df73a2c60256b7fd3b5f26/frame/frame.go#L665) of audio samples.

I think it should be:
```go
mid := int32((int64(l) + int64(r)) / 2)
```

instead of what is currently used for inter-channel decorrelation:
```go
mid := int32(math.Round((float64(l) + float64(r)) / 2.0)) // rounded up.
```

In other words, mid should be truncated, at least according to the [docs for correlation](https://github.com/mewkiz/flac/blob/3a6ac918b27457e9b2df73a2c60256b7fd3b5f26/frame/frame.go#L610).

---

@wader har du discord? : )

**Comment by wader** [comment]
> > What residual encoding mode is this? rice parameter or escape? some interesting diff in high/low or before/after zigzag? all from same partition?
> 
> The residuals are encoded using using rice parameter. There could for sure be issues in the high/low and/or zigzagging parts of encoding. Another aspect I think may be at fault (I think there may be more than one that's slightly broken) is the [inter-channel decorrelation](https://github.com/mewkiz/flac/blob/3a6ac918b27457e9b2df73a2c60256b7fd3b5f26/frame/frame.go#L665) of audio samples.
> 
> I think it should be:
> 
> ```go
> mid := int32((int64(l) + int64(r)) / 2)
> ```
> 
> instead of what is currently used for inter-channel decorrelation:
> 
> ```go
> mid := int32(math.Round((float64(l) + float64(r)) / 2.0)) // rounded up.
> ```

Aha i see, yeah float seems fishy. The div ends up with a float with a representation on wrong "side" when rounding up or so?

> In other words, mid should be truncated, at least according to the [docs for correlation](https://github.com/mewkiz/flac/blob/3a6ac918b27457e9b2df73a2c60256b7fd3b5f26/frame/frame.go#L610).

Nice explanation in the comment, remember thinking about that shift/truncate and why it works.

> @wader har du discord? : )

Yeap men har dålig koll på hur de fungerar, jag tror mitt username är "wader1". Hänger mest i diverse jqlang kanaler :)

**Comment by MarkKremer** [comment]
For `44 - 8-channel surround, 192kHz, 24 bit, using only 32nd order predictors.flac`, `WriteUnary` receives numbers above 255 but the current implementation doesn't support it. Possible fix:

```golang
func WriteUnary(bw *bitio.Writer, x uint64) error {
	for ; x > 8; x -= 8 {
		if err := bw.WriteByte(0x0); err != nil {
			return err
		}
	}

	bits := uint64(1)
	n := byte(x + 1)
	if err := bw.WriteBits(bits, n); err != nil {
		return err
	}
	return nil
}
```

Test:
```golang
package bits_test

import (
	"bytes"
	"github.com/icza/bitio"
	"github.com/mewkiz/flac/internal/bits"
	"testing"
)

func Test_Unary(t *testing.T) {
	w := new(bytes.Buffer)
	bw := bitio.NewWriter(w)

	var want uint64
	for ; want < 1000; want++ {
		// Write unary
		if err := bits.WriteUnary(bw, want); err != nil {
			t.Fatalf("error writing unary: %v", err)
		}
		// Flush buffer
		if err := bw.Close(); err != nil {
			t.Fatalf("error closing the buffer: %v", err)
		}

		// Read written unary
		r := bits.NewReader(w)
		got, err := r.ReadUnary()
		if err != nil {
			t.Fatalf("error reading unary: %v", err)
		}

		if got != want {
			t.Fatalf("the written and read unary doesn't match the original. got: %v, expected: %v", got, want)
		}
	}
}
```

This should fix the test for `44 - 8-channel surround, 192kHz, 24 bit, using only 32nd order predictors.flac` except for the difference in how the samplerate is encoded.

**Comment by mewmew** [comment]
With the Unary fix in #68 by @MarkKremer, this PR is now ready to be merged.

Very happy that we finally get support for LPC encoding. The original issue (#35) was created about 5 years ago : )

Cheerful regards,
Henry & Robin

**Comment by wader** [comment]
🥳

**Comment by mewmew** [comment]
While this PR has been merged, I just realized there is [one final TODO](https://github.com/mewkiz/flac/blob/b3cd2a5e8ac12c49dfa3dd3089f4e07d6540d3b4/encode_subframe.go#L25) we need to resolve before tagging the 1.0.10 release.

**Comment by wader** [comment]
Looking at the code it feels like it should be possible to postpone the wasted shift to just before write of each sample?

**Comment by mewmew** [comment]
> Looking at the code it feels like it should be possible to postpone the wasted shift to just before write of each sample?

I think the main issue is that the [sample is used](https://github.com/mewkiz/flac/blob/b3cd2a5e8ac12c49dfa3dd3089f4e07d6540d3b4/encode_subframe.go#L358) in `getLPCResiduals` to compute the residual, and that is based on already having wasted those bits. We could update the code to waste the sample bits in each place where a sample is read (e.g. `encodeConstantSamples`, `encodeVerbatimSamples`, etc).

For now, I settled on the defer method. While not the prettiest, it does help us with separation of concern, so we only handle wasted bits in one place (ref: https://github.com/mewkiz/flac/commit/97aacbbb937cfe23fdadcfbc97caf26793450aa9).

**Comment by mewmew** [comment]
@karlek, @MarkKremer and @wader, if you feel satisfied with the current state of affairs after the merge of this PR, then we can tag the 1.0.10 release : )
